# STAT 207 Group Lab Assignment 14 - [10 total points]

## Regularization Models & Selecting a Tuning Parameter

<hr>

## <u>Lab Grading</u>:

Should we grade your submission?  If not, write the netID of the submission to be graded.  (Note: We will only grade one assignment per group, and we'll pick the first one that says we should grade that submission.  We will assign the same grade to all team members.)

*For example*, you might respond: **grade this submission** or **my submission is under netID jdeeke**

mdm16

If you said **my submission is under netID** above, we will not read any more of your lab submission.

If you said **grade this submission** above, who worked with you on this submission?  Write both their **names** and **netIDs**.  Also, discuss and record your plans for break.

## <u>Purpose</u>:
You should work in groups of 2-3 on this report (not working in groups without permission will result in a point deduction). The purpose of this group lab assignment is to fit a regularized model and assess its performance on new data.
<hr>

## <u>Assignment Instructions</u>:

### Group Roles

Suggested and specified roles are provided below: 

#### Groups of 2

* **Driver**: This student will type the report.  While typing the report, you may be the one who is selecting the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  Specific ways to help may include: outlining the general steps needed to solve a question (providing the overview), locating examples within the course notes, and reviewing each line of code as it is typed.

#### Groups of 3

* **Driver**: This student will type the report.  They may also be the one to select the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  They may select the general approach to answering the question and/or a few steps to be completed along the way. 
* **Communicator**: This student will review the report (as it is typed) to ensure that it is clear and concise.  This student may also locate relevant examples within the course notes that may help complete the assignment.

<hr>

### Imports

In [3]:
#Run this
import pandas as pd                    # imports pandas and calls the imported version 'pd'
import matplotlib.pyplot as plt        # imports the package and calls it 'plt'
import seaborn as sns                  # imports the seaborn package with the imported name 'sns'
sns.set()  

## Case Study: Avoiding Underwater Weighing

We will look at data collected on 252 males in 1985.  In particular, we will consider a measure of the percent of body fat in these males, as measured by Siri's equation (`siri`).  This particular measure is obtained through an underwater weighing technique and is quite extensive and resource-intensive.  We would like to consider alternative (and easier to capture information) to get roughly the same information that the `siri` variable currently contains.  We have other body measures for these males available, including:

- **age**: Age (yrs)
- **weight**: Weight (lbs)
- **height**: Height (inches)
- **adipos**: BMI index
- **neck**: Neck circumference (cm)
- **chest**: Chest circumference (cm)
- **abdom**: Abdomen circumference (cm)
- **hip**: Hip circumference (cm)
- **dthigh**: Thigh circumference (cm)
- **knee**: Knee circumference (cm)
- **ankle**: Ankle circumference (cm)
- **biceps**: Extended bicepts circumference (cm)
- **forearm**: Forearm circumference (cm)
- **wrist**: Wrist circumference (cm)

**We will use all variables in the data as predictors except for the siri, brozek, density, and free (fat free weight) variables**, since these four variables are challenging measurements to obtain and therefore might not be available for a different male that is not included in our current data.

The code cell below will read in the data for you.  Be sure to run the cell. 

In [4]:
df = pd.read_csv('fat.csv')

### 1. [1 point] Select a Regularization Technique 

We work for a company that is measuring the proportion of body fat for customers in order to design suits that can fit well for a customer that is ordering the suit over the internet.  We know that asking these customers to perform many body measurements will be burdensome for the customers, so we'd like to streamline the process by only asking the customer to provide a few key measurements.  We can then use our models to select an appropriately cut and tailored suit for the customer.

**a)** For this situation, what would our primary purpose of fitting the model be?  In other words, are we concerned about making predictions or understanding structures? 

The primary purpose is making predictions. We want to use a few easy customer measurements to predict body fat well enough to choose an appropriately fitting suit, rather than mainly trying to understand the structure of the relationship between all body measurements and body fat.

**b)** We would like to use a regularization model for our fitted model.  What regularization technique would you suggest should be used based on the optimal design for the company and customer?

I would suggest using Lasso regression because it can shrink some slope coefficients all the way to 0. That helps select a smaller set of key measurements, which matches the company’s goal of asking customers for only a few body measurements.







### 2. [2.5 points] Prepare the Data

**a)** Split your data into training and test data.  Use a random state (you can choose your random state), and set aside 20% of your data for the test data.

In [5]:
from sklearn.model_selection import train_test_split

df_train, df_test = train_test_split(df, test_size=0.2, random_state=207)


**b)** We know that regularization models require some preparation of the data before the model can be fit.  Scale your $X$ explanatory variables for this model.  Remember that we don't want to use siri, brozek, desnity, or free as explanatory variables.

*Note:* You should perform this scaling in two stages.  You should both fit a scaler and transform your training data in the first stage.  In the second stage, you should adjust your test data by applying the scaler trained by your training data to your test data.

In [6]:
from sklearn.preprocessing import StandardScaler

X_train = df_train.drop(columns=['siri', 'brozek', 'density', 'free'])
X_test = df_test.drop(columns=['siri', 'brozek', 'density', 'free'])

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


**c)** Prepare your y variable for the model.

In [7]:
y_train = df_train['siri']
y_test = df_test['siri']


### 3. [3 points] Fitting a Regularized Model

**a)** We'd like to fit a model using the regularization technique suggested in Question 1.  Start by fitting your model with a $\lambda$ of 2 to your training data.

In [8]:
from sklearn.linear_model import Lasso

lasso_2 = Lasso(alpha=2, max_iter=10000)
lasso_2.fit(X_train_scaled, y_train)


Lasso(alpha=2, max_iter=10000)

**b)** How many of your slopes were equal to 0 for this model?  Which variables would you suggest should be retained in the model?

In [11]:
coef_2 = pd.Series(lasso_2.coef_, index=X_train.columns)


print("Number of slopes equal to 0:", (coef_2 == 0).sum())


Number of slopes equal to 0: 13


13

**c)** Fit another version of your regularized model to your training data, but this time using a $\lambda$ of 0.5.

In [13]:
lasso_5 = Lasso(alpha=0.5, max_iter=10000)
lasso_5.fit(X_train_scaled, y_train)


Lasso(alpha=0.5, max_iter=10000)

**d)** How many of your slopes were equal to 0 for this model?  Which variables would you suggest should be retained from this version of the model?

In [14]:
coef= pd.Series(lasso_5.coef_, index=X_train.columns)


print("Number of slopes equal to 0:", (coef == 0).sum())

Number of slopes equal to 0: 10


10

### 4. [2 points] Comparing Model Results

**a)** Apply each of your two models fit from **3a** and **3c** to your test set.  Calculate the $R^2$ on your test set for each of your models.

In [15]:
from sklearn.metrics import r2_score

y_pred_2 = lasso_2.predict(X_test_scaled)
y_pred_05 = lasso_05.predict(X_test_scaled)

r2_2 = r2_score(y_test, y_pred_2)
r2_05 = r2_score(y_test, y_pred_05)

print("R^2 for lambda = 2:", r2_2)
print("R^2 for lambda = 0.5:", r2_05)


R^2 for lambda = 2: 0.6395145301839442
R^2 for lambda = 0.5: 0.7456212635738764


**b)** Which $\lambda$ would you suggest based on your test data?

I would suggest using λ = 0.5 because it had the higher test data R^2 value. The model with λ = 0.5 had an R^2 of about 0.746, while the model with λ = 2 had an R^2 of about 0.640, so λ = 0.5 performed better on the test data.


**c)** Compare the results to another group (or two) in your lab.  Did your models from **3** include the same predictor variables?  Did you pick the same preferred $\lambda$ from our two choices in **4b**?  Did you achieve the same $R^2$ values on your test data from **4a**?  Did you have the same training/test data split from **2a**?

We compared with another group. Our models did not include the same predictor variables. We didchoose the same preferred λ value. Our test-set R^2 values were similar. We did not use the same training/test split, because our random states were different.


### 5. [1.5 points] A Limitation of the Results

Return to your selection of a regularization technique in Question 1 and recall that there are two primary purposes or benefits to using a regularization model.  Assess if there is still a concern in your model that the other regularization model approach could address.  Explain.

Lasso helps choose fewer variables, but it may drop predictors that are still useful, especially if measurements are highly related. Ridge could address this by keeping all predictors while shrinking their slopes.

### 6. AI Acknowledgement

Our course policy is that you should write all of your own interpretations and other narrative answers (phrases or sentences) yourself without the assistance of AI.  AI should not be used on Questions 1, 3b, 3d, and 4c of this assignment.  

You may use AI to help guide your code, although you should write all of your own code yourself (not copy-paste from another source) and you should cite your use of AI.  I would encourage you to try to generate any necessary code yourself first using course resources and using AI as a debugging tool if/when you reach an error that you can't figure out or to help you perform any coding tasks that are more advanced than we've demonstrated during class (intended only for projects).  

Did you use AI on this assignment?  Did you use other resources outside of our course-provided resources on this assignment?

none


If you used AI or other resources, answer the following questions to cite your usage.

- Which AI and/or resources did you use?
- What prompts did you ask it?
- How did you integrate the responses into your assignment?  Specifically, which questions or parts are associated with this usage?

Note: answering these three questions are enough for our course but may not be enough for a different course or context.

n/a